### DISCUSSION TITLE
Async support for Django Graphql view

### DISCUSSION THREAD

**Discussion by ccsv** [mirumee/ariadne#493] [Ideas]
As of Django 3.1 there is support for asynchronous views.

Can someone update the Django Graphql view to be asynchronous?

Also if I want my queries and mutations to be asynchronous what would be the best way of converting them? using the [@sync_to_async decorator](https://docs.djangoproject.com/en/3.1/topics/async/)?

**Comment by rafalp** [comment]
We plan to move `ariadne.contrib.django` into separate library with Ariadne 0.14 release. Hopefully this feature request will see more attention when that happens.

**Reply by markedwards** [reply]
I got an `AsyncGraphQLView` proof-of-concept working today.  Need to clean it up a bit, but I can probably do a PR next week.

**Reply by rafalp** [reply]
@markedwards please consider making it separate library instead as per my previous comment.

May be also worth it to get in touch with @kedlav who is planning to release separate Django-Ariadne library, but I don't know how things are on his side.

**Reply by markedwards** [reply]
Oh I thought you were only talking about the Relay work.  I was assuming by your response above you meant that Mirumee would continue to maintain contrib, just as a separate library.

**Reply by rafalp** [reply]
Nope. We want people to release all the amazing stuff that they are building without having to keep their work hostage to Ariadne release cycle. 😄

I'm open to adding "Amazing Ariadne Extras" page to either docs or main repo as sort of index to help with discoverability though.

**Reply by markedwards** [reply]
Got it, so the intention is separate packages for separate concerns.  `ariadne-django`, `ariadne-relay`, etc.

@kedlav what’s the state of your django project?  I’m happy to contribute if it helps.

**Reply by maxmorlocke** [reply]
We're neck deep in a pivot right now, I expect we'll be free to kick off on this within the next ~2 weeks, with the goal of having a package live on pypi by June 1.  We'll port over the existing logic in contrib, as well as add our existing pulls, some decorators, etc.

**Reply by rafalp** [reply]
Great to hear back from you @kedlav. I've figured that you are busy right now. Thank you for keeping us updated 👍

**Reply by markedwards** [reply]
@kedlav, how do you see async fitting in to the initial roadmap?  And do you need any help?

**Reply by maxmorlocke** [reply]
We don’t use any of django’s async ourselves, so I haven’t put any thought into it.  I think contributions from someone that can put the code through it’s paces in production sound fantastic.

**Reply by markedwards** [reply]
Thanks @kedlav.  We may or may not use async (I have a feeling we will), but I'm more than happy to contribute.

It's not particularly difficult to convert `ariadne.contrib.django.views.GraphQLView` to async.  I only had to change like 8 lines.  I think the only work here is devising a strategy to maximise code reuse between the two implementations (which shouldn't be super difficult).

@kedlav can you ping me when you have a repo going?  Thanks!

**Reply by markedwards** [reply]
@kedlav Another thing I've worked on a bit is a way of implementing the equivalent of Django's `@login_required` and `@permission_required` for an Ariadne resolver.  I'm not sure I've got a satisfactory solution yet, but I thought it might be in scope here.

**Reply by maxmorlocke** [reply]
We have a decorator we use for login_required.  We'll be pushing enhancements like this up into the repo as soon as we get through the 0.0.1 release (e.g. get it into the repo and building with existing code).  If you'd like to use ours, here's the source:

```
from rest_framework.exceptions import NotAuthenticated


def login_required():
    def wrapped_decorator(func):
        def wrapped(cls, info, *args, **kwargs):
            user = info.context["request"].user
            if not user.is_authenticated:
                raise NotAuthenticated()
            return func(cls, info, *args, **kwargs)

        return wrapped

    return wrapped_decorator
```

**Reply by maxmorlocke** [reply]
Alrighty, the initial port for ariadne_django is complete: 

- https://pypi.org/project/ariadne-django/0.1.0/#description
- https://github.com/reset-button/ariadne_django

I've included some issues to make it clear what Lexria is committing to build into the package, with milestones going over our initial targeted enhancements.  We're planning on porting our application to use ariadne_django in mid-April, but we welcome contributions and additional folks willing to help test.  Should be entirely compatible with ariadne 0.13's contrib.django package (barring pathing changes) - minor code organization changes have been applied.

Thanks again @rafalp and team.  Once Lexria's live with ariadne_django, I'll put in a pull request to the documentation site for ariadne to update the docs for 3rd party packages.

**Reply by markedwards** [reply]
Thanks @kedlav.  Regarding async, I think I have some ideas on how to do it properly.  Should I open an issue and plan to submit a PR for it?

**Reply by maxmorlocke** [reply]
Sounds good to me
